okhttp/src/jvmTest/java/okhttp3/RouteFailureTest.kt

```diff
@@ -0,0 +1,240 @@
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+
+@Flaky
+class RouteFailureTest {
+  private lateinit var socketFactory: SpecificHostSocketFactory
+  private lateinit var client: OkHttpClient
+
+  @RegisterExtension
+  val platform = PlatformRule()
+
+  @RegisterExtension
+  val clientTestRule = OkHttpClientTestRule()
+
+  private lateinit var server1: MockWebServer
+  private lateinit var server2: MockWebServer
+
+  private var listener = RecordingEventListener()
+
+  private val handshakeCertificates = localhost()
+
+  val dns = FakeDns()
+
+  val ipv4 = InetAddress.getByName("192.168.1.1")
+  val ipv6 = InetAddress.getByName("2606:2800:220:1:248:1893:25c8:1946")
+
+  val refusedStream = MockResponse()
+    .setHttp2ErrorCode(ErrorCode.REFUSED_STREAM.httpCode)
+    .setSocketPolicy(SocketPolicy.RESET_STREAM_AT_START)
+  val bodyResponse = MockResponse().setBody("body")
+
+  @BeforeEach
+  fun setUp(
+    server: MockWebServer,
+    @MockWebServerInstance("server2") server2: MockWebServer
+  ) {
+    this.server1 = server
+    this.server2 = server2
+
+    socketFactory = SpecificHostSocketFactory(InetSocketAddress(server.hostName, server.port))
+
+    client = clientTestRule.newClientBuilder()
+      .dns(dns)
+      .socketFactory(socketFactory)
+      .eventListenerFactory(clientTestRule.wrap(listener))
+      .build()
+  }
+
+  @Test
+  fun http2OneBadHostOneGoodNoRetryOnConnectionFailure() {
+    enableProtocol(Protocol.HTTP_2)
+
+    val request = Request(server1.url("/"))
+
+    server1.enqueue(refusedStream)
+    server2.enqueue(bodyResponse)
+
+    dns[server1.hostName] = listOf(ipv6, ipv4)
+    socketFactory[ipv6] = server1.inetSocketAddress
+    socketFactory[ipv4] = server2.inetSocketAddress
+
+    client = client.newBuilder()
+      .fastFallback(false)
+      .apply {
+        retryOnConnectionFailure = false
+      }
+      .build()
+
+    executeSynchronously(request)
+      .assertFailureMatches("stream was reset: REFUSED_STREAM")
+
+    assertThat(client.routeDatabase.failedRoutes).isEmpty()
+    assertThat(server1.requestCount).isEqualTo(1)
+    assertThat(server2.requestCount).isEqualTo(0)
+  }
+
+  @Test
+  fun http2OneBadHostOneGoodRetryOnConnectionFailure() {
+    enableProtocol(Protocol.HTTP_2)
+
+    val request = Request(server1.url("/"))
+
+    server1.enqueue(refusedStream)
+    server1.enqueue(refusedStream)
+    server2.enqueue(bodyResponse)
+
+    dns[server1.hostName] = listOf(ipv6, ipv4)
+    socketFactory[ipv6] = server1.inetSocketAddress
+    socketFactory[ipv4] = server2.inetSocketAddress
+
+    client = client.newBuilder()
+      .fastFallback(false)
+      .apply {
+        retryOnConnectionFailure = true
+      }
+      .build()
+
+    executeSynchronously(request)
+      .assertBody("body")
+
+    assertThat(client.routeDatabase.failedRoutes).isEmpty()
+    // TODO check if we expect a second request to server1, before attempting server2
```

Comment:
Yep

Reply:
(I think that’s the behavior we should test for, though we could create an issue and change that)